Source/Session.swift

```diff
@@ -556,7 +556,8 @@ extension Session: RequestDelegate {
         rootQueue.async {
             request.didCancel()
 
-            guard let task = self.requestTaskMap[request] else {
+            // Cancellation only has an effect on running or suspended tasks.
+            guard let task = self.requestTaskMap[request], [.running, .suspended].contains(task.state) else {
```

Comment:
I feel like we're trying to solve the problem with bandaids rather than addressing the core issue. The reason all these edge cases have been created is that `resume`, `suspend`, and `cancel` are not locking onto the same queue as the internal updates from the `Session`. The state transition is locked, but the execution of the matching operation isn't locked, and that produces the race. Why don't we just eliminate the race altogether?

```swift
@discardableResult
public func resume() -> Self {
    underlyingQueue.async {
        guard self.protectedMutableState.attemptToTransitionTo(.resumed) else { return }
        self.delegate?.resumeRequest(self)
    }

    return self
}
```

By switching `resume` to be called on the `underlyingQueue` (same as `rootQueue` in `Session`), you are now locking the state transition with the execution of the operation. This should solve the actual race and will allow you to remove these checks in the delegate as well as remove the async dispatch here at the beginning.

Reply:
Making the entire process asynchronous opens us up to host of other other issues. Namely:
- User expectations are no longer met. When a `URLSessionTask` has `cancel()` called, the `state` of that task is updated immediately, despite the fact that the delegate method indicating the actual cancellation will come later. It seems reasonable for users to expect the same behavior from Alamofire, especially since we don't offer any sort of async completion handler for these actions.
- In addition to user expectations not being met, internal expectations may no longer hold, as we could no longer rely on actions like cancellation being queryable immediately, meaning the library may do a lot of unnecessary work, like issuing a request retry, when it didn't need to. So the race is still there, we've just moved its effects somewhere else internally.

Ultimately, this is no longer a thread-safety issue but an order of operations issue. Even though we're using a serial queue internally, there's no way for us to guarantee the order of operations that have dependent effects when enqueueing things asynchronously. i.e. There's no way for us to insert a cancellation action before an already pending retry attempt. So since that problem isn't solvable (at least not without a huge amount of work) we aren't buying ourselves anything by putting the state transition on the same queue.

So while such a change might make dealing with task creation a bit easier, it would complicate the cancellation side of things quite a bit, and the state updates would no longer be immediately available to users.

Reply:
Seems like `underlyingQueue.sync` would solve the issue.

Reply:
Not really. `sync` just waits for the closure to completed. That means we'd still have cancellation enqueued after internal actions which query for that state. In addition, we'd block the calling queue, which may not be a trivial amount of time considering the `rootQueue` is also the `URLSession`'s `delegateQueue`, as well as the `underlyingQueue` for every `Request` created by that `Session` instance. It would also mean we couldn't call any of this API internally ourselves, as we'd deadlock.

I'll give a solution here more thought.

Reply:
It opens up a bigger issue of the queuing system design. This should be a change that we could make, but I'm not sure we can without hanging up this call in certain situations (i.e. root queue is parsing a big JSON blob). By using the same underlying queue across the board, I can't see how we could prevent this from happening. This is why we've used separate queues in the past.

Reply:
Based on offline Slack conversation, let's move forward with this as is right now.